### ISSUE TITLE
LSUV/Ortho with PReLU

### ISSUE THREAD

**Issue by ibmua** [ducha-aiki/caffenet-benchmark#17]
I've copied the lsuv.py script into my current caffe installation (not the one in here) and when I try to use it with PReLU units I get errors.

For LSUV:

```
Loading
conv_1 (128, 1, 5, 5)
conv_1 var =  0.0857792 mean =  0.19733
conv_1 var =  1.16208 mean =  0.189732
conv_1 var =  0.733266 mean =  0.193065
conv_1 var =  1.08708 mean =  0.190994
conv_1 var =  0.938934 mean =  0.191936
conv_1 var =  0.968886 mean =  0.191879
conv_1 var =  0.914341 mean =  0.192624
conv_1 var =  1.04143 mean =  0.191765
conv_1 var =  1.00926 mean =  0.191769
conv_1_rectifier (128,)
Traceback (most recent call last):
  File "/home/sharpy/caffe/tools/extra/lsuv.py", line 73, in <module>
    solver.net.forward(end=k)
  File "/home/sharpy/caffe/tools/extra/../../python/caffe/pycaffe.py", line 124, in _Net_forward
    return {out: self.blobs[out].data for out in outputs}
  File "/home/sharpy/caffe/tools/extra/../../python/caffe/pycaffe.py", line 124, in <dictcomp>
    return {out: self.blobs[out].data for out in outputs}
KeyError: 'conv_1_rectifier'


```

-- line 73 in my copy is 
        if 'LSUV' in init_mode:
            if var_before_relu_if_inplace:
                solver.net.forward(end=k) #this one

For Ortho:

```

Loading
conv_1 (128, 1, 5, 5)
conv_1_rectifier (128,)
Traceback (most recent call last):
  File "/home/sharpy/caffe/tools/extra/lsuv.py", line 66, in <module>
    weights=svd_orthonormal(v[0].data[:].shape)
  File "/home/sharpy/caffe/tools/extra/lsuv.py", line 17, in svd_orthonormal
    raise RuntimeError("Only shapes of length 2 or more are supported.")
RuntimeError: Only shapes of length 2 or more are supported.

```

Line 66 is
        if 'Orthonormal' in init_mode:
            weights=svd_orthonormal(v[0].data[:].shape) #it's this one

Meanwhile, I don't have any errors if I replace PReLU activation with ELU, or ReLU.

**Comment by ducha-aiki** [comment]
@ibmua the reason is that script tries to do init for the PReLU layer (since it has learnable parameters), but fails because of shape mismatch.
It is not the best practice, but usually I just add to lsuv_init.py after line
`for k,v in solver.net.params.iteritems():`
something like 
         try:
            print(k, v[0].data.shape)
        except:
            print('Error with layer',k, 'skipping it')
            continue
        if ('BN' in k) or ('bn' in k) or ('relu' in k ):
            print('Skipping layer')
            continue;
and add "relu" or "bn" to the name and tops of all layers with parameters, which are not "InnerProduct", nor "Convolution".

**Comment by ibmua** [comment]
That's what you did when benchmarking, right?
https://github.com/ducha-aiki/caffenet-benchmark

Why not put an if/else case into the lsuv.py script?

**Comment by ibmua** [comment]
Side note, is it meaningful to readjust the network from time to time during learning with LSUV to keep it from exploding?

**Comment by ducha-aiki** [comment]
@ibmua no, it will hurt performance. 
If properly initialized network explodes, that smth wrong with different things, i.e. too high learning rate. 
Also I have experienced this with ELUs in very deep network, fixed by setting alpha=0.5
Regarding your comment about adding skips - skip  "bn" is already there. I have added the relu one right now, but it wouldn`t help you :) I mean, "conv_1_rectifier" does not contain "relu" string.
